### ISSUE TITLE
Optimization using Tensorflow issue

### ISSUE THREAD

**Issue by ashishsme14** [tensorflow/tensorflow#46319]
We had given below test case, we written all steps as per instructions. nit sure when we went wrong. Please guide us -

We will look at three scenarios here -

- Tuning different optimisation algorithms, 
- Tuning learning rate and momentum of SGD optimizer
- Tuning beta values of Adam optimizer

```
import numpy as np
from sklearn.model_selection import GridSearchCV
from keras.models import Sequential
from keras.layers import Dense
from keras.wrappers.scikit_learn import KerasClassifier
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder,StandardScaler
from sklearn.utils import shuffle
from keras.utils.np_utils import to_categorical
from keras.optimizers import SGD,Adam
from matplotlib import pyplot
import seaborn as sns
import pandas as pd
from keras.models import model_from_json

```

- Load the Dataset
- Load the iris dataset using load_iris() function.
- Store the data of the iris dataset in the variable X.
- Store the target of the iris dataset in the variable y.
- Convert the variable y into categorial variable using function to_categorial and save it in variable y.
- Set seed value as 7 in the variable seed and use random.seed function in numpy to set seed value.
- Now shuffle the data X and y using shuffle function and save it in variables X ,Y.

```
iris = load_iris()
X=  iris.data
y = iris.target
y = to_categorical(iris.target,3)
seed = 7
np.random.seed(seed)
X, Y =  shuffle(X, y)
```

In variable optimizer pass the following optimizers as a list -
SGD,RMSprop,Adam,Nadam
In param_grid pass parameter optimizer as optimizer using dict

```
optimizer = ['SGD', 'RMSprop',  'Adam', 'Nadam']
param_grid = dict(optimizer=optimizer)
```

Create a sequential model
The model expects rows of data with 4 variables (the input_dim=4 argument)
The first hidden layer has 64 nodes and uses the relu activation function.
The second hidden layer has 32 nodes and uses the relu activation function.
The third hidden layer has 16 nodes and uses the relu activation function.
The output layer has 3 nodes and uses the softmax activation function.
While comipling the model pass the following parameters -
     -optimizer as optimizer
     -loss as categorical cross entropy 
     -metrics as accuracy.
Return the compiled model

```
def create_model(optimizer='adam'):
    model = Sequential()
    model.add(Dense(4, input_dim=4))
    model.add(Dense(64, activation='relu'))
    model.add(Dense(32, activation='relu'))
    model.add(Dense(16, activation='relu'))
    model.add(Dense(3, activation='softmax'))
    model.compile(loss='categorical_crossentropy', optimizer=optimizer, metrics=['accuracy']) 
    return model
```

Use the KerasClassifier function to call the model function with following parameters -
build_fn as create_model
batch_size as 10
verbose as 0
epochs as 10
Save the above in the variable model

`model = KerasClassifier(build_fn=create_model, batch_size=10, verbose=0, epochs=10)`

In grid use the GridSearchCV function and pass the following parameters -
estimator as model
param_grid as param_grid
n_jobs as 1
Now fit the model with X and Y using grid and save it in grid_result

```
grid = GridSearchCV(estimator=model, param_grid=param_grid, n_jobs=-1)
grid_result = grid.fit(X, Y)
```

```
print("Best: %f using %s" % (grid_result.best_score_, grid_result.best_params_))
means = grid_result.cv_results_['mean_test_score']
stds = grid_result.cv_results_['std_test_score']
params = grid_result.cv_results_['params']
for mean, stdev, param in zip(means, stds, params):
    print("%f (%f) with: %r" % (mean, stdev, param))
```

Best: 0.946667 using {'optimizer': 'RMSprop'}
0.753333 (0.082192) with: {'optimizer': 'SGD'}
0.946667 (0.047140) with: {'optimizer': 'RMSprop'}
0.913333 (0.083799) with: {'optimizer': 'Adam'}
0.866667 (0.106249) with: {'optimizer': 'Nadam'}

```
pyplot.figure(figsize=(10,8))
#pyplot.xticks(grid_result1.cv_results_['mean_test_score'])
pyplot.title("Performance metrics of each optimiser")
plot=sns.barplot(grid_result.cv_results_['mean_test_score'],optimizer)
pyplot.show() 
```

In variable learn_rate pass the following learn rates as a list -
0.001, 0.01,0.3
In variable momentum pass the following momentums as a list -
0.0, 0.4, 0.9

```
learn_rate = [0.001, 0.01, 0.3]
momentum = [0.0, 0.4, 0.9]
```

Use the same model parameters as above to construct the model in the function create_model1
In the variable optimizer pass the following parameters using the optimizer SGD
 -lr as learn_rate
 -momentum as momentum
While comipling the model pass the following parameters -
 -optimizer as optimizer
 -loss as categorical cross entropy 
 -metrics as accuracy.
Return the compiled model

```
def create_model1(learn_rate=0.01, momentum=0):
    model1 = Sequential()
    model1.add(Dense(4, input_dim=4))
    model1.add(Dense(64, activation='relu'))
    model1.add(Dense(32, activation='relu'))
    model1.add(Dense(16, activation='relu'))
    model1.add(Dense(3, activation='softmax'))
    optimizer = SGD(lr=learn_rate, momentum=momentum)
    model1.compile(loss='categorical_crossentropy', optimizer=optimizer, metrics=['accuracy'])    
    return model1
```

`model1 = KerasClassifier(build_fn=create_model1, batch_size=10, verbose=0, epochs=10)`

```
param_grid1 = dict(learn_rate=learn_rate,momentum=momentum )
grid1 = GridSearchCV(estimator=model1, param_grid=param_grid1, n_jobs=1)
```

Use random.seed function in numpy to set seed value.
Now fit the model with X and Y using grid1 and save it in grid_result1

```
np.random.seed(seed)
grid_result1 = grid1.fit(X, Y)
```

```
print("Best: %f using %s" % (grid_result1.best_score_, grid_result1.best_params_))
means1 = grid_result1.cv_results_['mean_test_score']
stds1 = grid_result1.cv_results_['std_test_score']
params1 = grid_result1.cv_results_['params']
for mean, stdev, param in zip(means1, stds1, params1):
    print("%f (%f) with: %r" % (mean, stdev, param))
```

Best: 0.893333 using {'learn_rate': 0.01, 'momentum': 0.9}
0.360000 (0.032660) with: {'learn_rate': 0.001, 'momentum': 0.0}
0.666667 (0.037712) with: {'learn_rate': 0.001, 'momentum': 0.4}
0.780000 (0.081650) with: {'learn_rate': 0.001, 'momentum': 0.9}
0.760000 (0.016330) with: {'learn_rate': 0.01, 'momentum': 0.0}
0.866667 (0.160278) with: {'learn_rate': 0.01, 'momentum': 0.4}
0.893333 (0.067987) with: {'learn_rate': 0.01, 'momentum': 0.9}
0.413333 (0.188562) with: {'learn_rate': 0.3, 'momentum': 0.0}
0.333333 (0.018856) with: {'learn_rate': 0.3, 'momentum': 0.4}
0.333333 (0.018856) with: {'learn_rate': 0.3, 'momentum': 0.9}

```
params1=pd.DataFrame(params1)
pyplot.figure(figsize=(10,8))
#pyplot.xticks(grid_result1.cv_results_['mean_test_score'])
pyplot.title("Performance metrics of SGD optimiser with different learning rates and momentum")
plot1=sns.barplot(params1["learn_rate"],grid_result1.cv_results_['mean_test_score'],hue=params1["momentum"])
plot1.set(ylabel='Score')
pyplot.show() 
```

```
beta_1 = [0.001, 0.01, 0.3]
beta_2 = [0.0, 0.4, 0.9] 
```

```
def create_model2(beta_1=0.01, beta_2=0):
    model2 = Sequential()
    model2.add(Dense(4, input_dim=4))
    model2.add(Dense(64, activation='relu'))
    model2.add(Dense(32, activation='relu'))
    model2.add(Dense(16, activation='relu'))
    model2.add(Dense(3, activation='softmax'))
    optimizer = Adam(beta_1=beta_1,beta_2=beta_2)
    model2.compile(loss='categorical_crossentropy', optimizer=optimizer, metrics=['accuracy'])
    return model2
```

`model2 =  KerasClassifier(build_fn=create_model2,batch_size=10, verbose=0,  epochs=10)`

```
param_grid2 = dict(beta_1=beta_1, beta_2=beta_2)
grid2 = GridSearchCV(estimator=model2, param_grid=param_grid2, n_jobs=1)
```

Use random.seed function in numpy to set seed value.
Now fit the model with X and Y using grid2 and save it in grid_result2

```
np.random.seed(seed)
grid_result2 = grid2.fit(X, Y)
```

```
print("Best: %f using %s" % (grid_result2.best_score_, grid_result2.best_params_))
means2 = grid_result2.cv_results_['mean_test_score']
stds2 = grid_result2.cv_results_['std_test_score']
params2 = grid_result2.cv_results_['params']
for mean, stdev, param in zip(means2, stds2, params2):
    print("%f (%f) with: %r" % (mean, stdev, param))
```

Best: 0.966667 using {'beta_2': 0.4, 'beta_1': 0.001}
0.526667 (0.264491) with: {'beta_2': 0.0, 'beta_1': 0.001}
0.966667 (0.009428) with: {'beta_2': 0.4, 'beta_1': 0.001}
0.840000 (0.133666) with: {'beta_2': 0.9, 'beta_1': 0.001}
0.320000 (0.032660) with: {'beta_2': 0.0, 'beta_1': 0.01}
0.813333 (0.111156) with: {'beta_2': 0.4, 'beta_1': 0.01}
0.766667 (0.151731) with: {'beta_2': 0.9, 'beta_1': 0.01}
0.360000 (0.032660) with: {'beta_2': 0.0, 'beta_1': 0.3}
0.926667 (0.041096) with: {'beta_2': 0.4, 'beta_1': 0.3}
0.766667 (0.151731) with: {'beta_2': 0.9, 'beta_1': 0.3}

```
params2=pd.DataFrame(params2)
pyplot.figure(figsize=(10,8))
#pyplot.xticks(grid_result1.cv_results_['mean_test_score'])
pyplot.title("Performance metrics of SGD optimiser with different beta values")
plot2=sns.barplot(params2["beta_1"],grid_result2.cv_results_['mean_test_score'],hue=params2["beta_2"])
plot2.set(ylabel='Score')
pyplot.show() 
```

```
with open("score.txt","w") as f:
    f.write(str(round(grid_result.best_score_,2)))
with open("params.txt","w") as f:
    f.write(str(grid_result.best_params_))

with open("score1.txt","w") as f:
    f.write(str(round(grid_result1.best_score_,2)))
with open("params1.txt","w") as f:
    f.write(str(grid_result1.best_params_))
    
with open("score2.txt","w") as f:
    f.write(str(round(grid_result2.best_score_,2)))
with open("params2.txt","w") as f:
    f.write(str(grid_result2.best_params_))

```        

```
def save_model(model):
    # saving model
    json_model = model.to_json()
    open('model.json', 'w').write(json_model)
    # saving weights
    model.save_weights('model.h5', overwrite=True)
classifier=create_model()
save_model(classifier)
```

**Comment by amahendrakar** [comment]
@ashishsme14,
This question is better asked on [StackOverflow](https://stackoverflow.com/questions/tagged/tensorflow) since it is not a TensorFlow bug or feature request. There is also a larger community that reads questions there. Thanks!

**Comment by ashishsme14** [comment]
please help us

**Comment by mihaimaruseac** [comment]
This question is better asked on StackOverflow as it asks for help in solving a particular problem. As such, it should not be here.
